Code/Editor/Lib/Tests/test_AzAssetBrowserRequestHandler.cpp

```diff
@@ -0,0 +1,653 @@
+        entries.push_back(SourceAssetBrowserEntry::GetSourceByUuid(m_uuidOfSource2));
+        Utils::ToMimeData(&mimeData, entries);
+
+        QDropEvent dropEvent(
+            QPoint(0, 0), Qt::DropAction::CopyAction, &mimeData, Qt::MouseButton::LeftButton, Qt::KeyboardModifier::NoModifier);
+
+        // when asset of type m_assetTypeOfModel is dragged, spawn a s_MockEditorComponentTypeId.
+        // when asset of type m_assetTypeOfActor is dragged, spawn a s_MockEditorComponentTypeId also - but high priority
+        testing::NiceMock<AssetTypeInfoBusHandlerMock> mockHandler1;
+        testing::NiceMock<AssetTypeInfoBusHandlerMock> mockHandler2;
+
+        // the actor handler should have a higher priority, so it should always 'win'
+        mockHandler1.InstallHandlersFor(AZ::Uuid(MockEditorComponent::s_MockEditorComponentTypeId), m_assetTypeOfModel);
+        mockHandler2.InstallHandlersFor(AZ::Uuid(MockEditorComponent::s_MockEditorComponentTypeId), m_assetTypeOfActor, 10);
+
+        DragAndDropEventsBus::Event(EditorViewport, &DragAndDropEvents::Drop, &dropEvent, ctx);
+
+        mockHandler1.BusDisconnect();
+        mockHandler2.BusDisconnect();
+
+        EXPECT_TRUE(dropEvent.isAccepted());
+        EXPECT_FALSE(m_createdEntities.empty());
+
+        // inspect the created entities:
+        ASSERT_EQ(2, m_createdEntities.size());
+
+        // One entity should have the first source, subid 2
+        // The other entity should have the second source, subid 123
+        // The order of entities is not defined, so we can't make assumptions.
+        // So put the expected values in an unordered_set, and pluck them out as they are found.
+        // it will be an error if we find one not in the set, or if there are still remainders after
+        // we've checked all entities.
+        AZStd::unordered_set<AZ::Data::AssetId> expectedAssetIds;
+        expectedAssetIds.insert(AZ::Data::AssetId(m_uuidOfSource1, 2));
+        expectedAssetIds.insert(AZ::Data::AssetId(m_uuidOfSource2, 123));
+        
+        for (auto entity : m_createdEntities)
+        {
+            auto allComponents = entity->FindComponents<MockEditorComponent>();
+            ASSERT_EQ(1, allComponents.size()); // exactly 1 component per entty
+            MockEditorComponent* mockComponent = allComponents[0];
+            ASSERT_NE(nullptr, mockComponent);
+
+            AZ::Data::AssetId currentAsset = mockComponent->m_primaryAssetSet;
+
+            ASSERT_TRUE(expectedAssetIds.find(currentAsset) != expectedAssetIds.end());
+            expectedAssetIds.erase(currentAsset);
+        }
+        EXPECT_TRUE(expectedAssetIds.empty()) << "Did not find all expected assets!";
+    }
+
+    // this test checks to make sure that when custom handlers are installed, and eat the event, the default
+    // processing does not happen:
+    TEST_F(AzAssetBrowserRequestHandlerFixture, DragEnterTest_DragSource_WithDefaultBehavior_OverridableByHandlers)
```

Comment:
I could add a test.  Not here though.  There's another problem with exported EBUSES which means that a bus that has a listener which exists in EditorLib (A dll), EditorLib.Tests (A dll), and AzToolsFramework (a staticlib) has linkage problems and can't acutally be instantiated in the test dll.  

There are ways around it but it gets/looks messy.  What the test above does is ensure that the default handler does not eat events that are already eaten (by anyone, including high priority handlers).

Note that higher priority handlers do not have to spawn entities.  Nor do they have to eat the event.  They can do anything wit hthe information.  So from a systems test, the only test I'd like to perhaps do there is ensure that they are called.  But that'd be re-testing the machinery of ebus itself, since its an EBUS call.  

Any suggestions here?  Basically, I can't currently put a listener to the drop bus anywhere in the EditorLIb.tests dll for linkage reasons.

Reply:
The reasoning makes sense. Since the existing test already ensures that the default handler doesn't consume the same events, I don't think explicitly verifying a custom handler is needed.